### ISSUE TITLE
Unit of measure for metrics depth ?

### ISSUE THREAD

**Issue by avanmalleghem** [DepthAnything/Depth-Anything-V2#165]
Hello,

I adapted the `run_video.py` you provide to integrate `metric_depth`. Here is the code I use. I mainly add some logs, provide a way to specify an output fps and add text on output with depth output.

```
import argparse
import cv2
import glob
import matplotlib
import numpy as np
import os
import torch
import timeit

from depth_anything_v2.dpt import DepthAnythingV2

def create_text(true_depth, image, frame_width, frame_height):
    steps = 20
    for y in range(1, steps):
        for x in range(1, steps):

            px_h = int(y * frame_height/steps)
            px_w = int(x * frame_width/steps)

            # Format the depth value (you can customize precision here)
            label_text = f'{true_depth[px_h, px_w]:.1f}'
            
            # Draw the text on the gradient image
            font = cv2.FONT_HERSHEY_SIMPLEX
            font_scale = 0.4
            font_thickness = 1
            text_color = (0, 0, 0)  # Black text
            text_position = (px_w, px_h)  # Position to the right of the color bar
            cv2.circle(image, (px_w, px_h), radius=5, color=(0, 0, 255), thickness=-1)
            cv2.putText(image, label_text, text_position, font, font_scale, text_color, font_thickness, cv2.LINE_AA)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Depth Anything V2 Metric Depth Estimation')
    
    parser.add_argument('--video-path', type=str)
    parser.add_argument('--input-size', type=int, default=518)
    parser.add_argument('--outdir', type=str, default='./vis_video_depth')
    
    parser.add_argument('--encoder', type=str, default='vitl', choices=['vits', 'vitb', 'vitl', 'vitg'])
    parser.add_argument('--load-from', type=str, default='checkpoints/depth_anything_v2_metric_hypersim_vitl.pth')
    parser.add_argument('--max-depth', type=float, default=80)
    
    parser.add_argument('--pred-only', dest='pred_only', action='store_true', help='only display the prediction')
    parser.add_argument('--grayscale', dest='grayscale', action='store_true', help='do not apply colorful palette')
    
    args = parser.parse_args()
    
    DEVICE = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'

    print(DEVICE)
    
    model_configs = {
        'vits': {'encoder': 'vits', 'features': 64, 'out_channels': [48, 96, 192, 384]},
        'vitb': {'encoder': 'vitb', 'features': 128, 'out_channels': [96, 192, 384, 768]},
        'vitl': {'encoder': 'vitl', 'features': 256, 'out_channels': [256, 512, 1024, 1024]},
        'vitg': {'encoder': 'vitg', 'features': 384, 'out_channels': [1536, 1536, 1536, 1536]}
    }

    depth_anything = DepthAnythingV2(**{**model_configs[args.encoder], 'max_depth': args.max_depth})
    depth_anything.load_state_dict(torch.load(args.load_from, map_location='cpu'))
    depth_anything = depth_anything.to(DEVICE).eval()
    
    if os.path.isfile(args.video_path):
        if args.video_path.endswith('txt'):
            with open(args.video_path, 'r') as f:
                lines = f.read().splitlines()
        else:
            filenames = [args.video_path]
    else:
        filenames = glob.glob(os.path.join(args.video_path, '**/*'), recursive=True)
    
    os.makedirs(args.outdir, exist_ok=True)
    
    margin_width = 50
    cmap = matplotlib.colormaps.get_cmap('Spectral')
    
    for k, filename in enumerate(filenames):
        start = timeit.default_timer()
        print(f'Progress {k+1}/{len(filenames)}: {filename}')
        
        raw_video = cv2.VideoCapture(filename)
        frame_width, frame_height = int(raw_video.get(cv2.CAP_PROP_FRAME_WIDTH)), int(raw_video.get(cv2.CAP_PROP_FRAME_HEIGHT))
        frame_rate = int(raw_video.get(cv2.CAP_PROP_FPS))
        print(f"width : {frame_width} - height : {frame_height} - rate : {frame_rate}")
        
        if args.pred_only: 
            output_width = frame_width
        else: 
            output_width = frame_width * 2 + margin_width
        
        output_path = os.path.join(args.outdir, os.path.splitext(os.path.basename(filename))[0] + '.mp4')
        out = cv2.VideoWriter(output_path, cv2.VideoWriter_fourcc(*"mp4v"), frame_rate, (output_width, frame_height))
        
        desired_output_fps = 1.0
        frame_count = 0
        while raw_video.isOpened():
            ret, raw_frame = raw_video.read()
            if not ret:
                break
            
            if frame_count % (frame_rate / desired_output_fps) == 0:
                true_depth = depth_anything.infer_image(raw_frame, args.input_size)
                
            depth = (true_depth - true_depth.min()) / (true_depth.max() - true_depth.min()) * 255.0
            depth = depth.astype(np.uint8)
            
            if args.grayscale:
                depth = np.repeat(depth[..., np.newaxis], 3, axis=-1)
            else:
                depth = (cmap(depth)[:, :, :3] * 255)[:, :, ::-1].astype(np.uint8)
            
            if args.pred_only:
                out.write(depth)
            else:
                split_region = np.ones((frame_height, margin_width, 3), dtype=np.uint8) * 255
                create_text(true_depth, depth, frame_width, frame_height)
                combined_frame = cv2.hconcat([raw_frame, split_region, depth])
                
                out.write(combined_frame)
                
            frame_count += 1
        
        raw_video.release()
        out.release()

        stop = timeit.default_timer()
        print('Time (seconds): ', stop - start)
```

I ran the following command : `python run_video.py --encoder vitl --load-from checkpoints/depth_anything_v2_metric_vkitti_vitl.pth --video-path ../assets/examples_video --outdir ./output-metrics`

And here is an example frame in output :

![image](https://github.com/user-attachments/assets/c3f81316-a708-4a46-aeca-c5187995c661)

My issue is : I don't understand the output values... I expected to have meters in output but it seems like decimeters ? Is it relevant ? Did I miss something ?

**Comment by jsckdon** [comment]
Do you now know what unit of measurement it is

**Comment by avanmalleghem** [comment]
Still waiting the answer to be honest...

**Comment by evidenciavisual** [comment]
just dropping by to remain a couple of things:
- The model predicts inverse depth (1 / depth). Therefore, smaller prediction corresponds to larger depth.
- To appropriately scale the predicted depth in your images, you might need to use ground control points with known distances (not sure with 16-bit raw output but the same apply)

I may be wrong but in general I would not expect actual magnitudes from depth prediction but a "normalized" data that should be scaled to case scenarios (i.e. some input in your code than accounts for a distance-range from actual closest to farthest values that your code should use to map/scale the depth data).

my two cents

**Comment by rayryeng** [comment]
> just dropping by to remind a couple of things:
> 
> * The model predicts inverse depth (1 / depth). Therefore, smaller prediction corresponds to larger depth.
> * To appropriately scale the predicted depth in your images, you might need to use ground control points with known distances (not sure with 16-bit raw output but the same apply)
> 
> I may be wrong but in general I would not expect actual magnitudes from depth prediction but a "normalized" data that should be scaled to case scenarios (i.e. some input in your code than accounts for a distance-range from actual closest to farthest values that your code should use to map/scale the depth data).
> 
> my two cents

The units of measure for **metric depth** are most definitely meters, as shown in the comment on inference in the README: https://github.com/DepthAnything/Depth-Anything-V2/blob/main/metric_depth/README.md?plain=1#L54

The above quote specifies relative depth, which has nothing to do with metric depth estimation.

For the issue of adapting this for video sequence, one method that I found that worked was to run the model on the first N frames, for example 32, to extract the metric depth, then use [Video Depth Anything](https://github.com/DepthAnything/Video-Depth-Anything) to extract the relative depth in a temporally consistent manner over the entire scene. We now have a source (relative depth) and target (metric depth), and we can find the scale and shift to add to the relative depth to match the metric depth by least squares optimization. You then take the scale and shift and apply it to all of the relative depth frames to achieve the final metric depth. However, the above comment does have merit where in order to match the magnitudes between them, we take the inverse of the metric depth to get "relative depth", find the scale and shift, then when you apply them to the relative depths from Video Depth Anything, you need to take the inverse again to take get the final metric depth.

This script from the [Metric Depth Video Toolbox](https://github.com/calledit/metric_depth_video_toolbox/blob/main/video_metric_convert.py) facilitates estimation the metric depth over a video sequence using the method I explained above.
